### PR TITLE
DAOS-2863 control: provide storage scan summary

### PR DESCRIPTION
**Pull request by tanabarr**
Report the total capacity and number of devices for SCM and
NVMe storage on each server in output of 'dmg storage scan', in
addition to the usual details. 'dmg storage scan --summary|-s' will
just report the summary.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

**Comment by tanabarr** [comment]
```
[tanabarr@wolf-71 daos_m]$ dmg storage query nvme-health -i
Active connections: [localhost:10001]
localhost:10001
        NVMe controllers and namespaces detail with health statistics:
                PCI Addr:0000:da:00.0 Serial:PHKS7505005J750BGN   Model:INTEL SSDPED1K750GA  Fwrev:E2010325 Socket:1
                        Namespace: id:1 capacity:750.00GB
                Health Stats:
                        Temperature:304K(31C)
                        Controller Busy Time:0s
                        Power Cycles:6
                        Power On Duration:2685h0m0s
                        Unsafe Shutdowns:4
                        Media Errors:0
                        Error Log Entries:0
                        Critical Warnings:
                                Temperature: OK
                                Available Spare: OK
                                Device Reliability: OK
                                Read Only: OK
                                Volatile Memory Backup: OK
                PCI Addr:0000:81:00.0 Serial:PHKS75050060750BGN   Model:INTEL SSDPED1K750GA  Fwrev:E2010325 Socket:1
                        Namespace: id:1 capacity:750.00GB
                Health Stats:
                        Temperature:304K(31C)
                        Controller Busy Time:0s
                        Power Cycles:6
                        Power On Duration:2685h0m0s
                        Unsafe Shutdowns:4
                        Media Errors:0
                        Error Log Entries:0
                        Critical Warnings:
                                Temperature: OK
                                Available Spare: OK
                                Device Reliability: OK
                                Read Only: OK
                                Volatile Memory Backup: OK
                PCI Addr:0000:87:00.0 Serial:CVFT5226001C1P6DGN   Model:INTEL SSDPEDMD016T4  Fwrev:8DV10171 Socket:1
                        Namespace: id:1 capacity:1.56TB
                Health Stats:
                        Temperature:296K(23C)
                        Controller Busy Time:10m0s
                        Power Cycles:178
                        Power On Duration:19085h0m0s
                        Unsafe Shutdowns:155
                        Media Errors:0
                        Error Log Entries:1
                        Critical Warnings:
                                Temperature: OK
                                Available Spare: OK
                                Device Reliability: OK
                                Read Only: OK
                                Volatile Memory Backup: OK
[tanabarr@wolf-71 daos_m]$ dmg storage scan -i
Active connections: [localhost:10001]
localhost:10001
        SCM Namespaces:
                Device:pmem1 Socket:1 Capacity:2.90TB
                Device:pmem0 Socket:0 Capacity:2.90TB
        NVMe controllers and namespaces:
                PCI Addr:0000:da:00.0 Serial:PHKS7505005J750BGN   Model:INTEL SSDPED1K750GA  Fwrev:E2010325 Socket:1
                        Namespace: id:1 capacity:750.00GB
                PCI Addr:0000:81:00.0 Serial:PHKS75050060750BGN   Model:INTEL SSDPED1K750GA  Fwrev:E2010325 Socket:1
                        Namespace: id:1 capacity:750.00GB
                PCI Addr:0000:87:00.0 Serial:CVFT5226001C1P6DGN   Model:INTEL SSDPEDMD016T4  Fwrev:8DV10171 Socket:1
                        Namespace: id:1 capacity:1.56TB
        Summary:
                SCM: 5.79TB in 2 devices
                NVMe: 3.03TB in 3 namespaces on 3 devices
[tanabarr@wolf-71 daos_m]$ dmg storage scan -i --summary
Active connections: [localhost:10001]
localhost:10001
        Summary:
                SCM: 5.79TB in 2 devices
                NVMe: 3.03TB in 3 namespaces on 3 devices
[tanabarr@wolf-71 daos_m]$ tanabarr/control-scan-summary

```

File: src/control/common/storage/pb_types.go [hunk 1]
```diff
@@ -130,6 +142,19 @@ func (ncs NvmeControllers) StringHealthStats() string {
 	return buf.String()
 }
 
+// Summary reports accumulated storage space and the number of controllers.
+func (ncs NvmeControllers) Summary() string {
```
Comment (mjmac) [inline src/control/common/storage/pb_types.go]: I'm really not sure I agree with setting this precedent of adding more UI-focused functionality to the types. I know I'm starting to sound like a broken record with this SOLID stuff, but let's step back a bit and look at what's going on in this PR:

We've got a net change of ~400LOC across 15 files in an effort to summarize storage output for the user. Any updates to the output format will likely need to touch a large subset of these files. Any changes to the storage types themselves will likely have a ripple effect across other files.

I really think we should be focusing on keeping the core types mostly "pure" and independent of any UI concerns. Stringers are useful for developer output and quick POC UIs, but they're not a good fit for finished systems.

I propose that the best way to approach this problem is to handle formatting and UI at the edge (i.e. on the outside of the client API). At the moment, this is just dmg. In the future, there could be other consumers of client API output, and those consumers will do their own slicing and dicing of the data.

Comment (tanabarr) [reply src/control/common/storage/pb_types.go]: these types are to be moved to the edge when we re-factor client NVMe, then the Stringer implementations will live in the client package

**Comment by tanabarr** [comment]
updated summary output
```
[tanabarr@wolf-71 daos_m]$ dmg -l wolf-71:10001 storage scan -i --summary
Active connections: [wolf-71:10001]
wolf-71:10001
    Summary:
        SCM: 5.79TB total capacity over 2 namespaces
        NVMe: 3.03TB total capacity over 3 controllers
```

File: src/control/common/storage/pb_types.go (hunk 1 above)
Comment (mjmac) [reply src/control/common/storage/pb_types.go]: I don't think that the point of my comment was addressed. It's all well and good to say that we're going to redo it later, but the more tech debt we accumulate the harder it is to change things later. We are quickly running out of time to "do it right later"... Later is almost here. Soon we'll be moving on to another phase of the contract work and will have less ability to focus on this stuff.

Comment (tanabarr) [reply src/control/common/storage/pb_types.go]: do you want a specific change or is it okay to revisit this when we stop using pb_types from within dmg?

Comment (mjmac) [reply src/control/common/storage/pb_types.go]: That's fine. We can tackle this later. I'm just sensitive to making more work for ourselves down the road with a bunch of small changes that add up to big painful upheavals later.

**Pull request merged**